Implement a builder-like constructor for `Target`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
In an oversight during #14228, we forgot to expose the `Target::new()` method to the rest of the crate. The following commits expose a Rust native `Target::new()` as well as some other methods which enable us to construct the `Target` using a builder-like architecture.

### Details
The following commits add a rust native constructor for the `Target` that uses a builder-like architecture to add all of the specified attributes in a more dynamic manner.

```rs
// To create a target with a specified description and number of qubits:
let sydney_target: Target = Target::new()
    .with_description("Sydney-Target")
    .with_num_qubits(4)
    .unwrap();

// To create a target with a specific dt value:
let seoul_target: Target = Target::new()
    .with_dt(10e-9)
    .unwrap();
```
The methods include the following:
* `new` - Rust native constructor, must be called before using any builder methods. Equivalent to `Default::default()`.
* `with_description` - To create a target with a specific description.
* `with_num_qubits` - (Fallible) Assigns an explicit number of qubits to the `Target`, may fail if the specified `qubit_properties` list length differs from the value of this number.
* `with_dt` - Assigns the `dt` value.
* `with_granularity` - Assigns a granularity value in units of `dt`.
* `with_min_length` - Assigns a minimal pulse length in units of `dt`.
* `with_pulse_alignment` - Assigns a gate instruction starting time.
* `with_acquire_alignment` - Assigns a measure instruction starting time.
* `with_qubit_properties` - (Fallible) Specifies the properties of each of the backend's qubits, if associated. May fail if the specified `qubit_properties` list length differs from the value of `num_qubits`.
* `with_concurrent_measurements` - Specifies which qubits need to be measured together.

raynelfss committed:
Inital: Expose `Target::new()`

raynelfss committed:
Inital: Expose `Target::new()`

jakelishman reviewed (commented):
Rust doesn't have keyword arguments, so a `new` method that's 8 `Option<T>`s long perhaps isn't the cleanest user interface.  Perhaps a builder-like interface would be cleaner?

```rust
let mut target = Target::new()
    .with_description("target")
    .with_num_qubits(4)
    .with_dt(10e-9);
```
or such?

raynelfss committed:
Add: Builder-like construction methods.
- Add detailed docstrings with warnings about where not to use them.

raynelfss committed:
Add: Builder-like construction methods.
- Add detailed docstrings with warnings about where not to use them.

raynelfss committed:
Add: Use builder-like construction in `py_new`

raynelfss committed:
Add: Use builder-like construction in `py_new`

raynelfss committed:
Fix: Separate result-handlling and panic versions of `num_qubit` and `qubit_properties` initializers.

raynelfss committed:
Fix: Separate result-handlling and panic versions of `num_qubit` and `qubit_properties` initializers.

raynelfss committed:
Test: Add rust unit tests.

raynelfss committed:
Test: Add rust unit tests.

raynelfss commented:
Out of curiosity, would it be easier to just have the constructor setter methods for `num_qubits` and `qubit_properties` (or have them together as one method), and allow the user to edit the rest of the values (`dt`, `minimal_length`, `granularity` `pulse_alignment`, `acquire_alignment` and `concurrent_measurements`)? Since all of these attributes are public and the only ones that are fallible are `num_qubits` and `qubit_properties` in respect to each other values, it would make sense to allow the users to do something like:

```rs
let mut target = Target::new().try_with_num_qubits(5)?.try_with_qubit_properties(vec![])?;
target.dt = Some(10e-9);
target.min_length = 2;
```

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -891,6 +895,381 @@ impl Target {
+    /// Adds a description to a newly constructed [Target].
+    ///
+    /// **Warning:** Should only be used during the construction process of [Target],
+    /// after calling [Target::new] or any other construction methods.
```
Is this actually necessary. If this is called later it's still valid right? Like it's not the normal pattern but I can't see anything wrong with doing:

```rust
let target = get_fully_populated_target_from_backend()

let updated_target = target.with_description("My modified target");
```

This applies to all the methods added here.

raynelfss replied:
Removed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// # Examples
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::Target;
````
The path here should be updated to:
```suggestion
    /// use qiskit_transpiler::target::Target;
```

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// /// # Examples
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::Target;
````
```suggestion
    /// use qiskit_transpiler::target::Target;
```

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// /// # Examples
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::Target;
````
```suggestion
    /// use qiskit_transpiler::target::Target;
```

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// # Examples
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::Target;
````
```suggestion
    /// use qiskit_transpiler::target::Target;
```

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// ```
+    pub fn try_with_num_qubits(mut self, num_qubits: usize) -> Result<Self, TargetError> {
+        if let Some(qubit_properties) = self.qubit_properties.as_ref() {
+            if num_qubits != qubit_properties.len() {
````
If this is required for the target to be valid this I'm not sure `qubit_properties` should be public. It seems easy to do:

```rust
let mut target = Target::new().with_num_qubits(100);
target.qubit_properties = Some(vec![py.None(), py.None()]);
```

Also we should move qubit properties to Rust, there isn't any reason to leave it in python. But this shouldn't be in this PR.

raynelfss replied:
Noted in #14306. Will make qubit properties private.

raynelfss replied:
Made private in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// let mut target = Target::new().with_min_length(3);
+    /// assert_eq!(target.min_length, 3)
+    /// ```
+    pub fn with_min_length(mut self, min_length: usize) -> Self {
````
Why is this `usize` while `granularity`,  `acquire_alignment`, and `pulse_alignment` are a `u32`?

raynelfss replied:
I think this slipped in when I first ported this to Rust, since I thought it represented a length within the Target and not something in units of `dt`. Should I update it in this PR?

mtreinish replied:
I think now or in #14248 (or the C api more generally) is the time to do it. Before 2.1.0rc1 is tagged at least. Not necessarily part of this pr though.

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// # Examples
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::Target;
````
```suggestion
    /// use qiskit_transpiler::target::Target;
```

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// # Examples
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::Target;
````
```suggestion
    /// use qiskit_transpiler::target::Target;
```

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// # Examples
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::Target;
````
```suggestion
    /// use qiskit_transpiler::target::Target;
```

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// # Examples
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::Target;
````
```suggestion
    /// use qiskit_transpiler::target::Target;
```

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// /// # Examples
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::Target;
````
```suggestion
    /// use qiskit_transpiler::target::Target;
```

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// let mut target = Target::new().with_qubit_properties(vec![]);
+    /// assert_eq!(target.num_qubits, Some(0));
+    /// ```
+    pub fn with_qubit_properties(self, qubit_properties: Vec<PyObject>) -> Self {
````
Is this actually a thing we should have in the rust builder api? It doesn't seem like it's something we'll ever define in Rust while it's a `PyObject`.

raynelfss replied:
Not until we have qubit properties in Rust. We may want to have this ready as a placeholder though.

mtreinish replied:
Do we have an issue tracking the move of qubit properties to rust?

raynelfss replied:
Yes #14306

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -891,6 +895,381 @@ impl Target {
+    ///        the length of this list must match the number of qubits in the target,
+    ///        where the index in the list matches the qubit number the properties
+    ///        are defined for. If some qubits don't have properties available you
+    ///        can set that entry to ``None``.
```
This is a bit confusing because this is `py.None()` not `Option::None`.

raynelfss replied:
Will change it to `py.None()`. This also probably means that we want `QubitProperties` to be an `Option` when stored in the `Target`.

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    /// // This should change the number of qubits to the length of the list.
+    /// assert_eq!(target.num_qubits, Some(0));
+    /// ```
+    pub fn try_with_qubit_properties(
````
We definitely shouldn't have `qubit_properties` as `pub` if we have this method. I did mention this before, but seeing it explicitly here we definitely need to remove the `pub` since it's unsound to have it mutable directly.

raynelfss replied:
Made private in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -891,6 +895,381 @@ impl Target {
+    ///        the length of this list must match the number of qubits in the target,
+    ///        where the index in the list matches the qubit number the properties
+    ///        are defined for. If some qubits don't have properties available you
+    ///        can set that entry to ``None``.
```
Same comment here about `None`.

raynelfss replied:
Fixed in 73228f5

mtreinish commented on crates/transpiler/src/target/mod.rs:
````diff
@@ -891,6 +895,381 @@ impl Target {
+    ///
+    /// ```rust
+    /// use qiskit_accelerate::target_transpiler::Target;
+    /// use qiskit_accelerate::nlayout::PhysicalQubit;
````
```suggestion
    /// use qiskit_transpiler::target::Target;
    /// use qiskit_circuit::PhysicalQubit;
```

raynelfss replied:
Fixed in 73228f5

raynelfss committed:
FIx: Address review comments

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
FIx: Address review comments

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -374,7 +378,7 @@ impl Target {
     ///
     /// Args:
     ///     operation (str): The operation name to get qargs for
-    /// Returns:
```
Is this correct? In sphinx-napoleon/google style which we use for qiskit it should be `Returns:`. See:https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html

raynelfss replied:
Will change it back to `Returns:`

mtreinish commented:
Given the timeframe and that this is a good to have internal usability thing I deferred this 2.2 to minimize the risk to 2.1.

raynelfss committed:
Chore: Use `Returns:` instead of `Returns` in docstrings

raynelfss committed:
Chore: Use `Returns:` instead of `Returns` in docstrings

Cryoris commented:
Basically the same reason to defer to 2.4 given that we're 2 days off RC1:

> Given the timeframe and that this is a good to have internal usability thing I deferred this 2.2 to minimize the risk to 2.1.

raynelfss committed:
Fix: Adapt to c changes

raynelfss committed:
Fix: Adapt to c changes

raynelfss committed:
Merge branch 'main' into target-expose-new

raynelfss committed:
Merge branch 'main' into target-expose-new

raynelfss committed:
Merge branch 'main' into target-expose-new

raynelfss committed:
Docs: Improve documentation

raynelfss committed:
Docs: Improve documentation

raynelfss committed:
Remove: Panicking methods in `Target`.
The following commit removed `with_num_qubits()`, `with_qubit_properties()`. in favor of the fallible methods.
Methods have been updated to use these additions.
